### PR TITLE
use hash to index Wikipedia titles

### PR DESCRIPTION
**Pull request by breunigs**
Indexing `wikipedia_article` titles using a btree takes ~380s, whereas using a hash brings this down to ~142s. The index statistics from `pg_stat_user_indexes` are nearly the same:

```
btree:
idx_scan   idx_tup_read   idx_tup_fetch
132460     17174          17174

hash:
idx_scan   idx_tup_read   idx_tup_fetch
132458     17198          17198
```

The `create_*_view.sql` are the ones making use of Wikipedia and are slightly faster, too. So it seems that at least for my area, faster index creation is not paid with slower lookups down the line.

The numbers for `wikipedia_redirects` go in the same direction, though I didn't look as closely there.

Both together reduce wall clock time from ~704s to ~463s, on top of some of the other patches.

### CODE REVIEW

**Comment by philippks** [comment]
This sounds good. On what area did you tested this? Do you know if this could have a negative impact when processing larger areas?

**Comment by breunigs** [comment]
Hamburg metropolitan area, or [8.9236 53.1336 10.8897 53.9682](http://bboxfinder.com/#53.133600,8.923600,53.968200,10.889700) to be exact. I added amenities+shops to the `mapping.yml`, though, so it should have some more entities. My machine won't be powerful enough for planet, but I could try all of Germany and see how that goes.

**Comment by breunigs** [comment]
I ran the change on germany.osm.pbf and the results are comparable. The full logs here are: https://gist.github.com/breunigs/f479685006cc4e6d50052280dd13ef6c I prepended each line with the duration it took from the previous one. This is a tad hard to read, so here's my interpretation:

```
# master.log
    0s 2023-03-12 17:49:06,375 - INFO - run pg_restore -j 2 --dbname osm --no-owner --section pre-data --section data /osmnames/data/import//wikipedia_article.sql.bin
    0s 2023-03-12 17:49:06,375 - INFO - run pg_restore -j 2 --dbname osm --no-owner --section pre-data --section data /osmnames/data/import//wikipedia_redirect.sql.bin
   12s 2023-03-12 17:49:18,219 - INFO - finished
   81s 2023-03-12 17:50:39,238 - INFO - finished
  362s 2023-03-12 17:56:41,116 - INFO - run wget --no-clobber --directory-prefix /osmnames/data/import/ https://download.geofabrik.de/europe/germany-latest.osm.pbf
```
so the import takes 12s and 12+81s respectively, as they run in parallel. The indexing starts once the import finishes, and that's the 362s. The wget is line is printed before it starts, so this is not the download (the file was already present anyway).

For comparison:
```
# optimize-wikipedia-index.log
    0s 2023-03-13 02:32:28,931 - INFO - run pg_restore -j 2 --dbname osm --no-owner --section pre-data --section data /osmnames/data/import//wikipedia_article.sql.bin
    0s 2023-03-13 02:32:28,932 - INFO - run pg_restore -j 2 --dbname osm --no-owner --section pre-data --section data /osmnames/data/import//wikipedia_redirect.sql.bin
   15s 2023-03-13 02:32:44,038 - INFO - finished
   79s 2023-03-13 02:34:02,647 - INFO - finished
  138s 2023-03-13 02:36:20,324 - INFO - run wget --no-clobber --directory-prefix /osmnames/data/import/ https://download.geofabrik.de/europe/germany-latest.osm.pbf
 ```
as expected, the index import takes roughly the same amount of time, but indexing is faster.

imposm import and other steps are all roughly the same, plus minus a couple of seconds. The next big difference is this:

```
# master.log
    0s 2023-03-12 19:24:48,640 - INFO - start executing sql file set_street_ids_by_street_name.sql
23715s 2023-03-13 02:00:03,474 - INFO - finished executing sql file set_street_ids_by_street_name.sql

# optimize-wikipedia-index.log
    0s 2023-03-13 04:03:47,352 - INFO - start executing sql file set_street_ids_by_street_name.sql
24031s 2023-03-13 10:44:18,824 - INFO - finished executing sql file set_street_ids_by_street_name.sql
```

Presumably because my backup ran from ~10:00 to ~10:45 for `optimize-wikipedia-index.log` made this one slower, as this part doesn't use Wikipedia at all. I failed to turn off backups apparently. I'll ignore this as an unfortunate mishap.

The last interesting change is the "materialize view" part:
```
# master.log
    6s 2023-03-13 02:01:06,017 - INFO - start executing sql file create_polygons_view.sql
    0s 2023-03-13 02:01:06,018 - INFO - start executing sql file create_points_view.sql
    0s 2023-03-13 02:01:06,019 - INFO - start executing sql file create_linestrings_view.sql
    0s 2023-03-13 02:01:06,019 - INFO - start executing sql file create_merged_linestrings_view.sql
    0s 2023-03-13 02:01:06,020 - INFO - start executing sql file create_housenumbers_view.sql
  162s 2023-03-13 02:03:48,459 - INFO - finished executing sql file create_housenumbers_view.sql
  270s 2023-03-13 02:08:18,367 - INFO - finished executing sql file create_polygons_view.sql
  188s 2023-03-13 02:11:26,418 - INFO - finished executing sql file create_points_view.sql
  917s 2023-03-13 02:26:43,079 - INFO - finished executing sql file create_linestrings_view.sql
  198s 2023-03-13 02:30:00,622 - INFO - finished executing sql file create_merged_linestrings_view.sql

# optimize-wikipedia-index.log
    6s 2023-03-13 10:45:52,893 - INFO - start executing sql file create_polygons_view.sql
    0s 2023-03-13 10:45:52,893 - INFO - start executing sql file create_points_view.sql
    0s 2023-03-13 10:45:52,893 - INFO - start executing sql file create_housenumbers_view.sql
    0s 2023-03-13 10:45:52,893 - INFO - start executing sql file create_merged_linestrings_view.sql
    0s 2023-03-13 10:45:52,893 - INFO - start executing sql file create_linestrings_view.sql
  308s 2023-03-13 10:51:01,244 - INFO - finished executing sql file create_housenumbers_view.sql
  140s 2023-03-13 10:53:21,627 - INFO - finished executing sql file create_polygons_view.sql
  188s 2023-03-13 10:56:29,906 - INFO - finished executing sql file create_points_view.sql
  877s 2023-03-13 11:11:06,586 - INFO - finished executing sql file create_linestrings_view.sql
  208s 2023-03-13 11:14:34,647 - INFO - finished executing sql file create_merged_linestrings_view.sql
```

The individual timings are all different, but their total time is the same -- 29mins. Given they all run in parallel, I assume it's due to chance and scheduling.

So to sum it up: indexing wikipedia is faster, everything else is the same, and `set_street_ids_by_street_name.sql` just ran at an unfortunate moment but is unrelated to this PR.

**Comment by philippks** [comment]
thanks for your investigation 💪 sounds good 👍
